src/models/ListRule.ts

```diff
@@ -16,32 +16,235 @@ limitations under the License.
+            return null;
+        }
+
+        const reason = content['reason'] || '<no reason>';
+        if (typeof reason !== "string") {
+            return null;
+        }
+
+        let type = event['type'];
+        let kind;
+        if (USER_RULE_TYPES.includes(type)) {
+            kind = EntityType.RULE_USER;
+        } else if (ROOM_RULE_TYPES.includes(type)) {
+            kind = EntityType.RULE_ROOM;
+        } else if (SERVER_RULE_TYPES.includes(type)) {
+            kind = EntityType.RULE_SERVER;
+        } else {
+            return null;
+        }
+
+        // From this point, we may need specific fields
+        if (RECOMMENDATION_BAN_VARIANTS.includes(recommendation)) {
+            return new ListRuleBan(entity, reason, kind);
+        } else if (RECOMMENDATION_OPINION_VARIANTS.includes(recommendation)) {
+            let opinion = parseInt(content['opinion'], 10);
+            return new ListRuleOpinion(entity, reason, kind, opinion);
+        } else {
+            // As long as the `recommendation` is defined, we expect
+            // that the rule is correct, just unknown.
+            return new ListRuleUnknown(entity, reason, kind, content);
+        }
+    }
+}
+
+/**
+ * A rule representing a "ban".
+ */
+export class ListRuleBan extends ListRule {
+    constructor(
+        /**
+         * The entity covered by this rule, e.g. a glob user ID, a room ID, a server domain.
+         */
+        entity: string,
+        /**
+         * A human-readable reason for this rule, for audit purposes.
+         */
+        reason: string,
+        /**
+         * The type of entity for this rule, e.g. user, server domain, etc.
+         */
+        kind: EntityType,
+    ) {
+        super(entity, reason, kind, Recommendation.Ban)
+    }
 }
+
+/**
+ * A rule representing an "opinion"
+ */
+export class ListRuleOpinion extends ListRule {
+    constructor(
+        /**
+         * The entity covered by this rule, e.g. a glob user ID, a room ID, a server domain.
+         */
+        entity: string,
+        /**
+         * A human-readable reason for this rule, for audit purposes.
+         */
+        reason: string,
+        /**
+         * The type of entity for this rule, e.g. user, server domain, etc.
+         */
+        kind: EntityType,
+        /**
+         * A number in [-100, +100] where -100 represents the worst possible opinion
+         * on the entity (e.g. toxic user or community) and +100 represents the best
+         * possible opinion on the entity (e.g. pillar of the community).
+         */
+        public readonly opinion: number
+    ) {
+        super(entity, reason, kind, Recommendation.Opinion);
+        if (!Number.isInteger(opinion)) {
+            throw new TypeError(`The opinion must be an integer, got ${opinion}`);
+        }
+        if (opinion < OPINION_MIN || opinion > OPINION_MAX) {
+            throw new TypeError(`The opinion must be within [-100, +100], got ${opinion}`);
+        }
```

Comment:
Do we also need to fix floating point numbers? `(Math.round(Math.max(OPINION_MIN, Math.min(opinion, OPINION_MAX)))`?

Reply:
We just checked two lines above that we're dealing with integers, so that feels redundant.

Reply:
oh ok